presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -248,11 +253,20 @@ public void enqueue(Lifespan lifespan, List<SerializedPage> pages)
         masterBuffer.addPages(serializedPageReferences);
 
         // process any pending reads from the client buffers
-        for (ClientBuffer clientBuffer : safeGetBuffersSnapshot()) {
+        List<ClientBuffer> buffers = safeGetBuffersSnapshot();
+        if (buffers.isEmpty()) {
+            return;
+        }
+        // handle potential for racy update of next index and client buffers present
+        int index = nextClientBufferIndex.get() % buffers.size();
+        for (int i = 0; i < buffers.size(); i++) {
+            buffers.get(index).loadPagesIfNecessary(masterBuffer);
+            index = (index + 1) % buffers.size();
             if (masterBuffer.isEmpty()) {
```

Comment:
This reordered checking if the masterBuffer is empty before loading the client. Depending on which case is more common this results in a bunch of extra lock acquisitions inside a client.

I think this also means that the index is incremented even if no data is loaded into it which might be undesirable?

Reply:
The code was a little cleaner this way (not a ton) but the assumptions are:
- this code path only matters when the producing side is slow, and the master buffer is usually emptied before all client buffers have been fed (or: most clients are waiting on data)
- since we just added data to the buffer, at least one iteration will receive data unless a race is lost in a very narrow window. Checking for empty first does not completely remove that window

If that occurs then yes, a client may “miss a turn”. I think that’s probably ok since missing a turn under these conditions implies that almost all clients are starved. The goal is not to be perfectly fair, but rather to spread the trickle of data more evenly.

Reply:
Could this introduce a weird odd/even bias? Or maybe not odd/even depending on how many clients a given input is divided up over, but some multiple of whatever clients buffer.

`        // handle potential for racy update of indexOffset and buffer list copy
        int index = nextClientBufferIndex.get() % buffers.size();
        for (int i = 0; i < buffers.size(); i++) {
            if (masterBuffer.isEmpty()) {
                break;
            }
            buffers.get(index).loadPagesIfNecessary(masterBuffer);
            index = (index + 1) % buffers.size();
        }
        // Resume from the next client buffer on the next iteration
        nextClientBufferIndex.set(index);
`

Doesn't look much worse to me and it does closer to what you would expect.

Reply:
Sure. I've reordered the check to the start of the loop to match your example with one change: the atomic write to `nextClientBufferIndex` is still only necessary when the loop exits early. If the loop completes then no atomic write is necessary because the next iteration starts from the same offset as the previous one.